src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexInterpreter.cs

```diff
@@ -300,13 +300,29 @@ private bool MatchRef(int index, int length, ReadOnlySpan<char> inputSpan, bool
             }
             else
             {
-                Debug.Assert(_textInfo != null, "If the pattern has backreferences and is IgnoreCase, then _textInfo must not be null.");
-                TextInfo ti = _textInfo;
                 while (c-- != 0)
                 {
-                    if (ti.ToLower(inputSpan[--cmpos]) != ti.ToLower(inputSpan[--pos]))
+                    char backreferenceChar = inputSpan[--cmpos];
+                    char currentChar = inputSpan[--pos];
+                    // If we are evaluating a backreference case-insensitive match, we first check if the characters at the position
+                    // are the same character.
+                    if (backreferenceChar != currentChar)
                     {
-                        return false;
+                        // If they are not the same character, then we need to check if the backreference character participates in case conversion
+                        // and if so, we need to fetch the case equivalences from our casing tables.
+                        Debug.Assert(_culture != null, "If the pattern has backreferences and is IgnoreCase, then _culture must not be null.");
+                        if (RegexCaseEquivalences.TryFindCaseEquivalencesForCharWithIBehavior(backreferenceChar, _culture, out ReadOnlySpan<char> equivalences))
+                        {
+                            // If the backreference character participates in case conversion, we use IndexOfAny to see if currentChar matches one of the
+                            // equivalences from our casing table.
+                            if (inputSpan.Slice(pos, 1).IndexOfAny(equivalences) == -1)
+                                return false;
+                        }
+                        else
+                        {
+                            // If the backreference character does not participate in case conversions we fail to match.
+                            return false;
+                        }
```

Comment:
Seems like the flow could be simplified a little to:
```C#
if (!RegexCaseEquivalences.TryFindCaseEquivalencesForCharWithIBehavior(backreferenceChar, _culture, out ReadOnlySpan<char> equivalences) ||
    equivalences.IndexOf(inputSpan[pos]) < 0)
{
    // The backreference character doesn't participate in case conversions, or it does but the input character
    // doesn't match any of its equivalents.  Either way, we fail to match.
    return false;
}
```
?